contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,477 @@
+
+        let data = Data {
+            destination_chain_id,
+            commands: messages
+                .into_iter()
+                .map(|msg| msg.try_into())
+                .collect::<Result<Vec<Command>, ContractError>>()?,
+        };
+        let encoded_data = data.encode();
+
+        let id = batch_id(&message_ids);
+        let msg_to_sign = msg_to_sign(&encoded_data);
+
+        Ok(Self {
+            id,
+            message_ids,
+            data,
+            msg_to_sign,
+        })
+    }
+}
+
+impl Proof {
+    pub fn new(
+        snapshot: Snapshot,
+        signers: HashMap<String, Signature>,
+        pub_keys: HashMap<String, multisig::types::PublicKey>,
+    ) -> Result<Proof, ContractError> {
+        let mut operators = snapshot
+            .participants
+            .iter()
+            .map(|(_, participant)| {
+                let pub_key = pub_keys
+                    .get(&participant.address.to_string())
+                    .ok_or(ContractError::PublicKeyNotFound {
+                        participant: participant.address.to_string(),
+                    })?
+                    .into();
+
+                Ok(Operator {
+                    address: evm_address(pub_key)?,
+                    weight: participant.weight.into(),
+                    signature: signers.get(participant.address.as_str()).cloned(),
+                })
+            })
+            .collect::<Result<Vec<Operator>, ContractError>>()?;
+        operators.sort_by(|a, b| a.address.cmp(&b.address));
+
+        Ok(Proof {
+            operators,
+            threshold: snapshot.quorum.into(),
+        })
+    }
+
+    pub fn encode(&self) -> HexBinary {
+        let mut addresses: Vec<Token> = Vec::new();
+        let mut weights: Vec<Token> = Vec::new();
+        let mut signatures: Vec<Token> = Vec::new();
+
+        self.operators.iter().for_each(|operator| {
+            addresses.push(Token::Address(ethereum_types::Address::from_slice(
+                operator.address.as_slice(),
+            )));
+            weights.push(Token::Uint(ethereum_types::U256::from_big_endian(
+                &operator.weight.to_be_bytes(),
+            )));
+
+            if let Some(signature) = &operator.signature {
+                signatures.push(Token::Bytes(signature.into()));
+            }
+        });
+
+        let threshold = Token::Uint(ethereum_types::U256::from_big_endian(
+            &self.threshold.to_be_bytes(),
+        ));
+
+        ethabi::encode(&[
+            Token::Array(addresses),
+            Token::Array(weights),
+            threshold,
+            Token::Array(signatures),
+        ])
+        .into()
+    }
+
+    pub fn encode_execute_data(&self, data: &Data) -> HexBinary {
+        ethabi::encode(&[
+            Token::Bytes(data.encode().into()),
+            Token::Bytes(self.encode().into()),
+        ])
+        .into()
+    }
+}
+
+impl Data {
```

Comment:
I added this here since here is where all logic related to evm encoding is. I ended up moving `Data` here instead since you also pointed out that destination_chain_id type may change depending of implementation. Thoughts?

Reply:
It's just awkward to have to navigate back-and-forth between files to see the logic of methods and the definition of the structure. I also don't think this `Data` structure and even the `CommandBatch` will be re-used by any other non-evm chain. Most if not all types defined in `types.rs` are evm-specific.